### ISSUE TITLE
[IDEA] Filter operator for inserting variables/references into text

### ISSUE THREAD

**Issue by saqimtiaz** [Jermolene/TiddlyWiki5#5121]
While we have a lot of text manipulation filter operators now, common uses cases like the following are still quite clunky:

`<span style={{{ [[color:]addsuffix{!!color_field}] }}}>`

I have a filter operator, which I've recently improved upon, which I previously did not think merited inclusion for the core. However, similar situations to the above example have come up very frequently in the last week and I am wondering if some variation thereof (with a better name) might be useful to add.

It operates on input titles and:

- all occurences of $(name)$ are replaced with variables of that name
- all occurences of $n$ are replaced with the nth operand, where n starts at 0.

`style={{{ [color:$0$;]printf{!!color}] }}}`

OR 

```
<$vars name=Gandalf age=1400>
<$text text={{{ [[My name is $(name)$ and $0$ my age is $(age)$ years.]printf[I think]] }}}/>
</$vars>
```

=>

`My name is Gandalf and I think my age is 1400 years`

This is an issue and not a PR as I am unsure if:
- such an operator would be easy to understand for a lay person
- is worth potential confusion it might cause with too many operators, considering that the problems it addresses can be solved in other ways.

Code:

```
exports.printf = function(source,operator,options) {
	var results = [];	
	function substitute(str) {		
		var output = str.replace(/\$\((\w+)\)\$/g, function(match,varname) {
			return options.widget.getVariable(varname,{defaultValue: ""})
		});
		output = output.replace(/\$(\d)\$/g, function(match,operandIndex) {
			if(operator.operands[operandIndex]) {
				return operator.operands[operandIndex];
			} else {
				return "$" + operandIndex + "$";
			}
		});
		return(output);
	}
	
	source(function(tiddler,title) {
		if(title) {
			results.push(substitute(title));
		}
	});
	return results;
};

})();
```

**Comment by Jermolene** [comment]
Awesome, makes a lot of sense. However, I worry that it might be inaccessible to people who are beginning with the filter syntax, and it introduces another language when we perhaps have too many already. I suggested elsewhere using backticks for attribute values that are to be wikified:

```
<$text text=`My name is <<name>> and my age is <<age>> years`>
```

However, I'm not wild about this idea:

* Any accidental wikitext in the string will mess things up
* Performance is relatively poor because there's no opportunity to cache the parse tree or widget tree
* Users are sometimes surprised that the results of macros/transclusions etc are themselves wikified

Maybe there's a hybrid where we introduce backticked attributes as a shorthand syntax for the variable substitution you've proposed.

**Comment by saqimtiaz** [comment]
@Jermolene We are on the same page in that:
- I think wikify should only be used when absolutely necessary due to the performance implications, and that we should try to lead people towards alternative code patterns rather than make it easier to overuse wikify.
- My filter implementation is not easy to grasp for non-developers. Which is why I've never made a PR for it. It was loosely inspired by the idea of re-using macro syntax and even existing macros within filters.

```
 <$text text=`My name is <<name>> and my age is <<age>> years`>
```

> Maybe there's a hybrid where we introduce backticked attributes as a shorthand syntax for the variable substitution you've proposed.

I think limiting it to just variable substitution is sensible and indeed easier to understand. However, I would favour a "variable substitution filter" in addition to or instead of the backticked attributes. The backticked attributes approach doesn't work when the base string is a derived from a variable or text reference.

`<$text text={{{ [<base-string>sub-variables[]] }}}/>`

where base-string might be: `My name is <<name>> and my age is <<age>> years`

Lastly note that in #4697 we have also discussed using backticks for escapable strings within filters, e.g: 

```
[join`\n`]
```

**Comment by kookma** [comment]
@saqimtiaz I do most of them using macros (text substitution) and filter operation. But I support your idea! I like the 

```
"string with vars".format(...)
```
or

```
"string with vars".printf(...)
```

I also understand Jeremy concerns to have the same philosophy as current filter operation/operators, but I think working with strings should be much simpler than what we have now!

**Comment by AnthonyMuscio** [comment]
#5110 Perhaps inspiring consideration of this it is certainly related.

[IDEA] Short cut to wikification #5110

**Comment by CodaCodr** [comment]
Seems there's a lot of co-thinking going on here - Saq, TonyM, Jeremy, me (https://groups.google.com/g/tiddlywikidev/c/KHQOFBzsbXQ/m/wVsTJ-eTCAAJ). If I left anyone out, sorry, I'm just catching up.

I'm not invested in backticks, but I don't see a workable alternative that so easily reduces the syntax noise of transcluded filters.

@saqimtiaz I think the wikify step should be included - but perhaps it needs to be "flagged" as a request? (Shame though - _more_ noise). Perhaps `` stuff `` ? (double BTs)

**Comment by CodaCodr** [comment]
@saqimtiaz @Jermolene 

> I think limiting it to just variable substitution is sensible and indeed easier to understand. 

I kind of agree. So to avoid wikification, if the core were to make currentTiddler's fields _preformed variables_ (much like macro args become `<<__arg__>>`), that would be one step closer to having "almost both":

```<span style=`<<my-prop>>:<<core-prepared-field-var>>` >my span</span>```

Unless, of course, there's a huge cost to doing that - I have some "data only" tiddlers with a ton of fields...

**Comment by twMat** [comment]
Semi-layman here, regarding how easily understood the syntax is:

```
<$vars name=Gandalf age=1400>
<$text text={{{ [[My name is $(name)$ and $0$ my age is $(age)$ years.]printf[I think]] }}}/>
</$vars>
```

Isn't that example showing *two* separate new syntaxes? One about evaluating `$(...)$` and the other concerning `$...$ .... printf[...]` ?

**$(...)$** - IMO this one is a very nice idea. We do already assume that people use/understand `$(...)$` in the context that this syntax normally/currently appears in, i.e for text substitution of a variable defined outside of the macro... so why would it *not* be OK in this new-but-very-similar context?

I don't quite understand in what context one would use the proposed printf parameter so I'll leave it at that.

**Comment by saqimtiaz** [comment]
> `$(...)$` - IMO this one is a very nice idea. We do already assume that people use/understand `$(...)$` in the context that this syntax normally/currently appears in, i.e for text substitution of a variable defined outside of the macro... so why would it _not_ be OK in this new-but-very-similar context?

1. Users currently do not need to understand text substitution in macros to use filters. So in the area of filters, that is new syntax.
2. You cannot just substitute variables in all filter operands. It would not be backwards compatible and, you would also need a way to escape the $(name)$ syntax for when you want to have that literal text in an operand.
3. Therefore you need a way to signify that variable substitution should happen, for example:
  a. by using backticks to denote such an operand.  (similar to operator[operand] but with variable substitution), 
  ```  
 operator`operand`
  ``` 
  b. or, by specifically calling a filter operator that does the variable substitution (call it sub-variables[] or print[] or... ).

**Comment by twMat** [comment]
>1. Users currently do not need to understand text substitution in macros to use filters. So in the area of filters, that is new syntax.

Yeah, my point is just that your syntax elegantly reuses an established syntax for a very similar concept so it should be minimally difficult to understand.

>[...] you need a way to signify that variable substitution should happen, for example:
>a. by using backticks to denote such an operand. (similar to operator[operand] but with variable substitution),

IMO putting out quote like characters to have something execute seems backward to me. HOWEVER, I would expect the "base case" to be that one *does want* the substitution to happen when using that syntax. So IMO it makes more sense with a signifier for when it should *not* happen. Ironically, this makes quote type characters, like backticks, suitable because such characters often signify that something is literal.

**Comment by saqimtiaz** [comment]
> IMO putting out quote like characters to have something execute seems backward to me. HOWEVER, I would expect the "base case" to be that one _does want_ the substitution to happen when using that syntax. So IMO it makes more sense with a signifier for when it should _not_ happen. Ironically, this makes quote type characters, like backticks, suitable because such characters often signify that something is literal.

Not backwards compatible.

**Comment by twMat** [comment]
>Not backwards compatible.

ah :-(

**Comment by saqimtiaz** [comment]
Closing in favour of #6663 which tries to address the same needs and has overlap with the ideas proposed here.
